wolfcrypt/src/port/atmel/atmel.c

```diff
@@ -896,13 +909,71 @@ int atcatls_verify_signature_cb(WOLFSSL* ssl, const byte* sig, unsigned int sigS
     return ret;
 }
 
+static int atcatls_set_certificates(WOLFSSL_CTX *ctx) 
+{
+    #ifndef ATCATLS_MAX_CERT_SIZE
+        #define ATCATLS_MAX_CERT_SIZE 560
+    #endif
+
+    int ret = 0;
+    ATCA_STATUS status;
+    size_t signerCertSize = ATCATLS_MAX_CERT_SIZE;
+    uint8_t signerCert[ATCATLS_MAX_CERT_SIZE];
+    size_t deviceCertSize = ATCATLS_MAX_CERT_SIZE;
+    uint8_t deviceCert[ATCATLS_MAX_CERT_SIZE];
+    int devPemSz, signerPemSz;
+    char devCertChain[2048];
+
+    /*Read signer cert*/
+    status = tng_atcacert_read_signer_cert((uint8_t*)&signerCert, &signerCertSize);
+    if (ATCA_SUCCESS != status) {
+        ret = atmel_ecc_translate_err(ret);
+        return ret;
+    }
+    /*Read device cert signed by the signer above*/
+    status = tng_atcacert_read_device_cert((uint8_t*)deviceCert, &deviceCertSize, (uint8_t*)signerCert);
+    if (ATCA_SUCCESS != status) {
+        ret = atmel_ecc_translate_err(ret);
+        return ret;
+    }
+    /*Generate a PEM chain of device certificate.*/
+    XMEMSET(devCertChain, 0, sizeof(devCertChain));
```

Comment:
Hi @vppillai,

This conversion to PEM is not needed. You can use `wolfSSL_CTX_use_certificate_chain_buffer_format` and provide the two certificates in a DER chain and specify `WOLFSSL_FILETYPE_ASN1`. The device cert should be first then the signer. That will avoid the 2048 buffer and conversion to PEM and back to DER (which happens in `wolfSSL_CTX_use_certificate_chain_buffer`).

Code would look like this:

```
size_t certSize;
uint8_t derChain[ATCATLS_MAX_CERT_SIZE * 2];
int derChainSz;

status = tng_atcacert_read_signer_cert(derChain, &certSize);
derChainSz = certSize;

status = tng_atcacert_read_device_cert(&derChain[certSize], &certSize, derChain);
derChainSz += certSize;

ret = wolfSSL_CTX_use_certificate_chain_buffer_format(ctx, derChain, derChainSz, WOLFSSL_FILETYPE_ASN1);
```

I did not test this, but should be close. That will save a lot of stack space and processing power.

David

Reply:
I tried this on my setup and negotiation is failing. I do not have the infra to debug it currently. 

You mentioned that device cert should be first then the signer. But in the snippet, the signer is packed into derChain first and then the device cert. So, I tried putting them into a packed struct with the device struct first using 521/522 sizes. Even in that setup, negotiation is failing.

Reply:
Yeah you'll have to read the signer then device, but the order will need to be swapped. Since the sizes are known it seems possible to read them in the right places directly. The ATCA library has these sizes you can get from headers. I will be able to test this out before we merge it. I'd like it to just use DER / ASN.1 without converting to PEM.